romancal/resample/resample.py

```diff
@@ -299,41 +307,44 @@ def resample_many_to_one(self):
 
         log.info("Resampling science data")
         members = []
-        for img in self.input_models:
-            inwht = resample_utils.build_driz_weight(
-                img,
-                weight_type=self.weight_type,
-                good_bits=self.good_bits,
-            )
-            if not hasattr(img.meta, "background"):
-                self._create_background_attribute(img)
-            blevel = img.meta.background.level
-            if not img.meta.background.subtracted and blevel is not None:
-                data = img.data - blevel
-            else:
-                data = img.data
-
-            xmin, xmax, ymin, ymax = resample_utils.resample_range(
-                data.shape, img.meta.wcs.bounding_box
-            )
+        with self.input_models:
+            for i, img in enumerate(self.input_models):
+                inwht = resample_utils.build_driz_weight(
+                    img,
+                    weight_type=self.weight_type,
+                    good_bits=self.good_bits,
+                )
+                if not hasattr(img.meta, "background"):
+                    self._create_background_attribute(img)
+                blevel = img.meta.background.level
+                if not img.meta.background.subtracted and blevel is not None:
+                    data = img.data - blevel
+                else:
+                    data = img.data
 
-            driz.add_image(
-                data,
-                img.meta.wcs,
-                inwht=inwht,
-                xmin=xmin,
-                xmax=xmax,
-                ymin=ymin,
-                ymax=ymax,
-            )
-            del data, inwht
-            members.append(str(img.meta.filename))
+                xmin, xmax, ymin, ymax = resample_utils.resample_range(
+                    data.shape, img.meta.wcs.bounding_box
+                )
 
-        members = (
-            members
-            if self.input_models.filepaths is None
-            else self.input_models.filepaths
-        )
+                driz.add_image(
+                    data,
+                    img.meta.wcs,
+                    inwht=inwht,
+                    xmin=xmin,
+                    xmax=xmax,
+                    ymin=ymin,
+                    ymax=ymax,
+                )
+                del data, inwht
+                members.append(str(img.meta.filename))
+                self.input_models.discard(i, img)
+
+        # FIXME: what are filepaths here?
+        # members = (
+        #     members
+        #     if self.input_models.filepaths is None
+        #     else self.input_models.filepaths
+        # )
```

Comment:
This is my fault.  There was a request to store the actual file names of the L2 input files (as opposed to their meta.filenames, which may become disconnected).  filepaths got populated with the file names from the association file in the model container here: https://github.com/spacetelescope/romancal/blob/e559890ee4679835b11e98e17f33ac819733176d/romancal/datamodels/container.py#L184

Reply:
Thanks for commenting on this. I don't yet fully understand how these are used or carried through the pipelines. Some discussion on expectations for this would be helpful. For example, when running the `mosaic` pipeline is the expectation that these will be the original input filenames? If so, I think this  code in skymatch will lose the filepaths:
https://github.com/spacetelescope/romancal/blob/e559890ee4679835b11e98e17f33ac819733176d/romancal/skymatch/skymatch_step.py#L100

Reply:
@schlafly I believe this is now handled in resample (instead of the container):
https://github.com/spacetelescope/romancal/pull/1241/files#diff-f87241f9a890ac31d6813379a09b6eb7e1052e3ab53039449a8f0df91aa030e1R351
but it would be great if this could be confirmed (I don't see any test failures but I'm not sure if there is a test for this feature).

Reply:
Thanks, sorry for losing this thread!  The issue was that meta.filename can easily become disconnected from the actual file name on disk, and there was a request for storing the actual file names on disk used to create a L3 file, rather than the meta.filenames that contributed.  My initial impression here
https://github.com/spacetelescope/romancal/pull/1241/files#diff-f87241f9a890ac31d6813379a09b6eb7e1052e3ab53039449a8f0df91aa030e1R351
is that this returns to the previous behavior of using meta.filename.  If there's a convenient way to ask the container for the original on-disk filenames, that would be best.

Reply:
Thanks. Is there a test for this (so I can poke it to see what's going on with and without this PR)?

To make sure I'm understanding the goal, let's say an association is provided to the mosaic pipeline that lists members "foo.asdf" and "bar.asdf" which have meta.filename attributes that are something else "chutes.asdf" and "ladders.asdf". The goal is to have the output resampled product contain "foo.asdf" and "bar.asdf". I think that should be easy to do by using the members info from `Library.asn` but I'll have to run some tests.

Reply:
@schlafly I switched to populating this using the asn members (which are the input filenames) in:
[449a58d](https://github.com/spacetelescope/romancal/pull/1241/commits/449a58d5f9d3029aaf151c46f581c77b38fc46db)
I didn't see any test differences so I'm assuming there is no test for this behavior (and the mosaic pipeline input files all have meta.filenames that match the real filenames).

I modified the test data for the mos pipeline test (locally) renaming one file in the association and on disk (but leaving it's meta.filename the same) and then run the pipeline. The result from resample recorded the modified name (the real filename) and not meta.filename.

I'd be happy to open an issue to track adding a test for this or adding one to this PR (I think it should be possible to add as a resample only unit test but that would require some investigation). Which would you prefer, an issue to track adding the test or adding it to this PR?

Reply:
Perfect, thanks, that looks great.  Just making sure I hadn't missed it before, that's a new feature in ModelLibrary that wasn't available to me when I wrote that bit of code before?  Great.  It looks like when the ModelLibrary is constructed from a list of models, these expnames default to meta.filename, but that seems like it isn't a regression since for some set of in-memory models I don't know what to do otherwise?

I tried to trace the code briefly and was a bit confused about what the resample step is now allowed to support as input.  The docstring is presumably out of date:
https://github.com/spacetelescope/romancal/blob/449a58d5f9d3029aaf151c46f581c77b38fc46db/romancal/resample/resample.py#L61-L62
This bit:
https://github.com/spacetelescope/romancal/blob/449a58d5f9d3029aaf151c46f581c77b38fc46db/romancal/resample/resample.py#L80-L84
says either a list of models or a ModelLibrary.
This bit
https://github.com/spacetelescope/romancal/blob/449a58d5f9d3029aaf151c46f581c77b38fc46db/romancal/resample/resample.py#L134-L135
suggests to me that it has to be a ModelLibrary (which presumably then always has the asn member).

I don't have strong preferences for an issue vs. an addition to this PR for a new test.  It looks like this PR is pretty close to being ready to merge, so I'm happy to delay a new test to an issue.  Thanks!

Reply:
Thanks!

> Just making sure I hadn't missed it before, that's a new feature in ModelLibrary that wasn't available to me when I wrote that bit of code before?

Exactly. I think the members entries previously got lost in skymatch due to:
https://github.com/spacetelescope/romancal/blob/747667895099a80ef283838a8499bac391cb1c46/romancal/skymatch/skymatch_step.py#L100

> It looks like when the ModelLibrary is constructed from a list of models, these expnames default to meta.filename, but that seems like it isn't a regression since for some set of in-memory models I don't know what to do otherwise?

Yup! Although we certainly could change that if something else seems more useful. The general idea is to make the `ModelLibrary` always have a "valid" association (even if it's made up based off the models) so the step code has a consistent interface (with `ModelContainer` I think you only get something in `asn_table` if the input is an association).

Re: resample step

Thanks for the catch. Yes the `ResampleData` class now only accepts a `ModelLibrary`. The `ResampleStep` should accept the same input as before (except for a `ModelLibrary` instead of a `ModelContainer`).
I pushed a change to the docstring in: https://github.com/spacetelescope/romancal/pull/1241/commits/6184692d1af66db786b59769598fc1ce6a346b97
and the error message in: https://github.com/spacetelescope/romancal/pull/1241/commits/899c9aa38bc275bc654d161571bca106cec56e67

I still have to rebase this to deal with the horde of conflicts following https://github.com/spacetelescope/romancal/pull/1314 so the CI is unhappy temporarily.

Reply:
I believe the linked commits addressed the documentation issues (thanks for finding those!) and I was able to "rebase". Rerunning the regtests shows the same results as linked in the above description (the run link is up-to-date).